### Tallfølger

Tallfølger er tall som kommer i rekkefølge etter hverandre. Noen tallfølger gis med en formel, og andre gis ved en slags algoritme.

```{admonition} Forutsetninger
:class: note
Denne siden forutsetter at man har vært gjennom delkapitlene
- {ref}`Variabler og syntaks`
- {ref}`Kalkulatorer og pakker`
- {ref}`Løkker`
- {ref}`Lister`
```

Her kommer noen vanlige tallfølger. Formlene under er gitt for tallene $n\in \{1, 2, 3, \dots\}$.

::::{grid} 3
:gutter: 1

:::{grid-item-card} Navn
:::

:::{grid-item-card} Formel
:::

:::{grid-item-card} Tall
:::

:::{grid-item-card}
De naturlige tallene
:::

:::{grid-item-card}
$n$
:::

:::{grid-item-card}
$\{1, 2, 3, 4, \dots\}$
:::


:::{grid-item-card}
Kvadrattallene
:::

:::{grid-item-card}
$n^2$
:::

:::{grid-item-card}
$\{1, 4, 9, 16, \dots\}$
:::

:::{grid-item-card}
Rektangeltallene
:::

:::{grid-item-card}
$n(n+1)$
:::

:::{grid-item-card}
$\{2, 6, 12, 20, \dots\}$
:::

:::{grid-item-card}
Trekanttallene
:::

:::{grid-item-card}
$\frac{n(n+1)}{2}$
:::

:::{grid-item-card}
$\{1, 3, 6, 10, \dots\}$
:::

::::

#### Programmere tallfølger med formel

Vi kan enkelt programmere tallfølger med formel. Da kan vi finne ulike summer. Vi kan også finne hvilke tall som er i følgen, og hvilke som ikke er det.

La oss finne summen av de $20$ første kvadrattallene og undersøke om tallet $289$ er et kvadrattall.

In [1]:
s = 0 # Summen starter som null

for n in range(1, 21): # For-løkken går fra 1 til én mindre enn den siste parameteren, altså 20.
    s += n**2          # Legger til summen. n**2 er formelen for kvadrattallene.
    print(n, n**2)     # Skriver ut indeks (plassering) og tall i tallfølgen.

print(f"Summen er {s}")

1 1
2 4
3 9
4 16
5 25
6 36
7 49
8 64
9 81
10 100
11 121
12 144
13 169
14 196
15 225
16 256
17 289
18 324
19 361
20 400
Summen er 2870


Som vi ser er tallet $289$ et kvadrattall, det er nummer 17 i følgen. I tillegg er summen av de $20$ første kvadrattallene lik $2870$.

#### Programmere tallfølger med algoritme

Det er ikke like lett å finne en formel for alle tallfølger. Noen ganger er det enklere å programmere tallfølgen fra en algoritme.

La oss se på *fibonaccitallene*. De er gitt ved denne algoritmen.

- Start med tallene $\{0, 1\}$
- Hvert neste tall er summen av de to forrige tallene.

La oss programmere starten på denne tallfølgen.

In [2]:
fib = [0, 1]

while len(fib) < 10:
    ny_fib = fib[-1] + fib[-2] # Tar summen av det siste og det nest siste tallet i listen.
    fib.append(ny_fib)         # Legger inn tallet i listen.

print(fib)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


Dette programmet legger inn det neste tallet i listen som summen av de to siste tallene så lenge listens lengde er under 10. Derfor skriver programmet ut de første ti fibonaccitallene.

---

#### Oppgaver

##### Oppgave 1

Hvis kvadrattallene er tallene med formel $n^2$ når $n\in \{1, 2, 3, \dots\}$ så må tallene med formel $n^3$ hete *kubikktallene*.

Skriv ut summen av de første $20$ kubikktallene og undersøk om $6859$ er et kubikktall.

##### Oppgave 2

Skriv ut de første 20 toerpotensene $2^n$. Har du sett disse tallene før?

##### Oppgave 3

> Se på koden for fibonaccitallene i eksempelet over.

*Det gylne snitt* er et viktig tall i matematikken. Hvis vi tar forholdet mellom et fibonaccitall og det forrige tallet, så vil verdien til dette forholdet bli mer og mer lik det gylne snitt desto lengre ut vi kommer i tallfølgen.

Lag et program som regner ut det gylne snitt ved å bruke fibonaccitallene.

```{admonition} Fasit: Det gylne snitt
:class: dropdown
Det gylne snitt er omtrent lik $\phi \approx 1.618033988749895$.
```

##### Oppgave 4

*Tribonaccitallene* er gitt ved følgende algoritme.

- Start med tallene $\{0, 1, 1\}$
- Det neste tallet er summen av de tre forrige tallene.

Lag et program som skriver ut de første 20 tribonaccitallene.

##### Oppgave 5

Primtallene kjenner du til, ikke sant? Primtallene er tallene som bare får et heltall som svar dersom man deler de på seg selv eller $1$.

Lag et program som skriver ut alle primtallene under $100$.

In [4]:
fib = [0, 1]

while len(fib) < 100:
    ny_fib = fib[-1] + fib[-2] # Tar summen av det siste og det nest siste tallet i listen.
    fib.append(ny_fib)         # Legger inn tallet i listen.
    print(fib[-1]/fib[-2])

print(fib)

1.0
2.0
1.5
1.6666666666666667
1.6
1.625
1.6153846153846154
1.619047619047619
1.6176470588235294
1.6181818181818182
1.6179775280898876
1.6180555555555556
1.6180257510729614
1.6180371352785146
1.618032786885246
1.618034447821682
1.6180338134001253
1.618034055727554
1.6180339631667064
1.6180339985218033
1.618033985017358
1.6180339901755971
1.618033988205325
1.618033988957902
1.6180339886704431
1.6180339887802426
1.618033988738303
1.6180339887543225
1.6180339887482036
1.6180339887505408
1.6180339887496482
1.618033988749989
1.618033988749859
1.6180339887499087
1.6180339887498896
1.618033988749897
1.618033988749894
1.6180339887498951
1.6180339887498947
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.618033988749895
1.